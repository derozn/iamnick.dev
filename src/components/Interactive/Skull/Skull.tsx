import React from 'react';
import { useLoader } from 'react-three-fiber';
import { GLTFLoader } from 'three/examples/jsm/loaders/GLTFLoader';
import { TextureLoader } from 'three';

import useStore from '#store';

const Skull = () => {
  const { gltfUrl, textureUrl } = useStore((state) => state.skull);

  const skullTexture = useLoader(TextureLoader, textureUrl);
  const skullGLTF = useLoader(GLTFLoader, gltfUrl);

  return (
    <group rotation={[0, -Math.PI / 4, 0]}>
      <mesh name="skull-head" {...skullGLTF.scene.children[3]}>
        <meshBasicMaterial attach="material" map={skullTexture} flatShading />
      </mesh>
      <mesh name="skull-jaw" {...skullGLTF.scene.children[2]}>
        <meshBasicMaterial attach="material" map={skullTexture} flatShading />
      </mesh>
    </group>
  );
};

export default Skull;
