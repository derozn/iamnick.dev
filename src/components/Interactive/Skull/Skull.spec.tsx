import React, { Suspense } from 'react';
import { create } from 'react-test-renderer';
import { useLoader } from 'react-three-fiber';

import useStore from '#store';

import Skull from './Skull';

jest.mock('#store');
jest.mock('react-three-fiber');

const mockUseStore = useStore as jest.Mock<any>; // Forgive me...
const mockUseLoader = useLoader as jest.Mock<any>; // And again...

// Will work on this..
describe('components/Skull', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders skull', () => {
    mockUseStore.mockReturnValue(() => ({
      skull: { gltfUrl: '/gltf-url', textureUrl: '/texture-url' },
    }));

    mockUseLoader
      .mockImplementationOnce(() => 'texture')
      .mockImplementationOnce(() => ({
        scene: { children: [undefined, undefined, { name: 'skull-jaw' }, { name: 'skull-head' }] },
      }));

    const wrapper = create(
      <Suspense fallback={null}>
        <Skull />
      </Suspense>,
    );
    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});
