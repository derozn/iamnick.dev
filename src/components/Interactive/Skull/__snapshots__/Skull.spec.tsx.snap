// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Skull renders skull 1`] = `
<group
  rotation={
    Array [
      0,
      -0.7853981633974483,
      0,
    ]
  }
>
  <mesh
    name="skull-head"
  >
    <meshBasicMaterial
      attach="material"
      flatShading={true}
      map="texture"
    />
  </mesh>
  <mesh
    name="skull-jaw"
  >
    <meshBasicMaterial
      attach="material"
      flatShading={true}
      map="texture"
    />
  </mesh>
</group>
`;
